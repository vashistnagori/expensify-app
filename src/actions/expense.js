import uuid from 'uuid';

export const addExpense=({description='', note='', amount=0, createdAt=''}={})=>{
    return {
        type:"addExpense", 
        expenseDetail:{id:uuid(), description, note, amount, createdAt}
    }
    }
    
export const removeExpense=({id=0}={})=>{
    return {
        type:"removeExpense", 
        id
    }
    }
    
export const editExpense=(id, updatedExpense)=>{
        return {
            type:"editExpense", 
            id,
            updatedExpense
        }
    }
