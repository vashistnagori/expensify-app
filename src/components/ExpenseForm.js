import React from 'react';
import moment from 'moment';
import {SingleDatePicker} from 'react-dates';
//import 'react-dates/lib/css/_datepicker.css'; now in app.js file

export default class ExpenseForm extends React.Component{
state={
                description: this.props.expense ?  this.props.expense.description : "",
                note: this.props.expense ?  this.props.expense.note : "",
                amount: this.props.expense ? ( this.props.expense.amount / 100).toString() : "",
                createdAt: this.props.expense ?  moment(this.props.expense.createdAt) : moment(),
                calanderFocused:false,
                error:"",
        };
 decriptionChange=(e)=>{
     const description= e.target.value;
     this.setState(()=>({description}));
 };
 noteChange=(e)=>{
    const note= e.target.value;
    this.setState(()=>({note}));
};
amountChange=(e)=>{
    const amount= e.target.value;
    if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){
        this.setState(()=>({amount}));
    }
};
dateChange=(createdAt)=>{
    if(createdAt){
        this.setState(()=>({createdAt}));
    }
}
forcusChange=({focused})=>{
this.setState(()=>({calanderFocused:focused}));
};
formSubmit=(e)=>{
    e.preventDefault();
    if(!this.state.description || !this.state.amount){
        this.setState(()=>({error:"please provide decription and amount"}));
    }
    else{
        this.setState(()=>({error:""}));
        this.props.onSubmit({
            description: this.state.description,
            amount:this.state.amount*100,
            createdAt:this.state.createdAt.valueOf(),
            note: this.state.note
        });
    }
};


    render(){
        return(<div>
           {this.state.error && <p>{this.state.error}</p> } 
            <form onSubmit={this.formSubmit}>
            <input 
            type='text'
            placeholder='description'
            value={this.state.description}
            onChange={this.decriptionChange}
            autoFocus
            />
            <input 
            type='number'
            placeholder='amount'
            value={this.state.amount}
            onChange={this.amountChange}
            />
            <SingleDatePicker
            date={this.state.createdAt} 
            onDateChange={this.dateChange} 
            focused={this.state.calanderFocused}
            onFocusChange={this.forcusChange} 
            numberOfMonths={1}
            isOutsideRange={()=>false}
            />
            <textarea
            placeholder="add a note here"
            value={this.state.note}
            onChange={this.noteChange}>
            </textarea>
            <button>Add Expense</button>
            </form>
            </div>);
    }
}