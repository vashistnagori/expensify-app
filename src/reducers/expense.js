
export const expense_reducer=(ps=[],action)=>{
    switch(action.type){
    case("addExpense"):
    return [...ps, action.expenseDetail];
    break;
    case("editExpense"):
    let updatedExpenseArray=ps.map((current)=>{
        if(current.id === action.id){
         return {
             ...current,
             ...action.updatedExpense
         }
        }
        else{
          return current;
        }
    });
    return updatedExpenseArray;
    break;
    case("removeExpense"):
    return ps.filter(o => o.id !== action.id);;
    break;
    case("SET_EXPENSE"):
    return action.expense;
    break;
    default:
    return ps;
    }
    };