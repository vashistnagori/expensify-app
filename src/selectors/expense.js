import moment from 'moment';

export const filterExpense=(exp, { text, SortBy, startDate, endDate})=>{
    const filtered_expences=exp.filter((c)=>{
        const createdAtMoment= moment(c.createdAt);
        const match_start_date= startDate ? startDate.isSameOrBefore(createdAtMoment, "day") : true ;
        const match_end_date= endDate ? endDate.isSameOrAfter(createdAtMoment, "day") : true;
        const match_text= text==="" || c.description.toLowerCase().indexOf(text.toLowerCase()) > -1;
    return match_text && match_start_date && match_end_date;
    }).sort((a,b)=>{
    if(SortBy=='date'){
        return a.createdAt>=b.createdAt? -1 : 1;
    }else if(SortBy=='amount'){
        return a.amount>=b.amount? -1: 1;
    }
    });
    return filtered_expences;
    }