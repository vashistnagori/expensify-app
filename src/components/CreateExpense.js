import React from "react";
import ExpenseForm from './ExpenseForm';
import {connect} from 'react-redux';
import {addExpense} from '../actions/expense';

export class CreateExpense extends React.Component{
  onSubmit=(expense)=>{
    this.props.dispatch_addexpenses(expense);
    this.props.history.push("/");
  }
  render(){
    return( <div>
      <h1>Add Expense</h1>
      <ExpenseForm onSubmit={this.onSubmit}/>
      </div>);
  }
}



    const mapDispatchToProps=(dispatch)=>({
        dispatch_addexpenses:(expense)=>dispatch(addExpense(expense))
    });

export default connect(undefined, mapDispatchToProps)(CreateExpense);
