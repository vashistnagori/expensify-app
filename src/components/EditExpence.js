import React from "react";
import {connect} from 'react-redux';
import ExpenseForm from './ExpenseForm';
import {editExpense, removeExpense, startRemoveExpense, startEditExpense} from '../actions/expense';

// export const EditExpence=(props)=>{
//    return(
//     <div>
//     <ExpenseForm 
//      expense={props.expense}  
//      onSubmit={(expense)=>{
//         props.dispatch_editExpense(props.expense.id, expense);
//         props.history.push("/");
//      }}/>
//      <button onClick={(e)=>{
//         props.dispatch_removeExpense({id:props.expense.id});
//         props.history.push("/");
//        }}>remove</button>
//     </div>
//     );
//    }

export class EditExpence extends React.Component{
   onSubmitfun=(expense)=>{
      this.props.dispatch_editExpense(this.props.expense.id, expense);
      this.props.history.push("/");
   }
   onDelFun=(e)=>{
      this.props.dispatch_removeExpense({id:this.props.expense.id});
      this.props.history.push("/");
     }

   render(){
      return(
             <div>
             <ExpenseForm 
              expense={this.props.expense}  
              onSubmit={this.onSubmitfun}/>
              <button onClick={this.onDelFun}>remove</button>
             </div>
             );
   }
}

const mapStateToProps=(state, props)=>{
   const expense= state.expenses.find((e)=>{
      return e.id === props.match.params.id;
   });
   if(expense){ return {expense}; } else{ return {};  }
}

const mapDispatchToProps=(dispatch)=>{
return{
   dispatch_removeExpense:({id})=>startRemoveExpense({id})(dispatch),
   dispatch_editExpense:(id, expense)=> startEditExpense(id, expense)(dispatch)
}
}


export default connect(mapStateToProps, mapDispatchToProps)(EditExpence);