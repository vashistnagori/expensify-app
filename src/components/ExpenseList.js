import React from 'react';
import {connect} from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import {filterExpense} from '../selectors/expense';


// const ExpenseList =(props)=>{
//     const expenseListItems= props.expenses.map((c)=>{
//        return (<div>
//         <ExpenseListItem key={"list"+c.id} {...c} />
//        </div>);
//     });
//     return (<div>{expenseListItems}</div>);
// };

export const ExpenseList =(props)=>(<div>
{
    props.expenses.length !== 0 ? (props.expenses.map((c)=>{
                                        return (<div>
                                        <ExpenseListItem key={"list"+c.id} {...c} />
                                        </div>);
                                        })
    ) : (<p>No Expenses</p>)
}
    </div>);


const mapStateToProps=(state)=>{
    return {
        expenses: filterExpense(state.expenses,state.filter)
    }
}

export default connect(mapStateToProps)(ExpenseList);
