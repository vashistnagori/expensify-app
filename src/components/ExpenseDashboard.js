import React from "react";
import ExpenseList from './ExpenseList';
import ExpenseListFilter from './ExpenseListFilter';

const ExpenseDashboard=(props)=>{
    //console.log(props);
   return(<div>
    <p>this is homepage</p>
    <ExpenseListFilter />
    <ExpenseList />
    </div>);
}
export default ExpenseDashboard;