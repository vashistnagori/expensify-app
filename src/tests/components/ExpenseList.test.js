import React from 'react';
import { shallow } from 'enzyme';
import { ExpenseList } from '../../components/ExpenseList';
import {testexp} from '../fixture/expense';

test("testing ExpenseList component with values",()=>{
const wrapper=shallow(<ExpenseList expenses={testexp} />);
expect(wrapper).toMatchSnapshot();
});

test("testing ExpenseList component with No values",()=>{
    const wrapper=shallow(<ExpenseList expenses={[]} />);
    expect(wrapper).toMatchSnapshot();
    });
    