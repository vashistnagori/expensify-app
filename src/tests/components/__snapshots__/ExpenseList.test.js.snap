// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing ExpenseList component with No values 1`] = `
<div>
  <p>
    No Expenses
  </p>
</div>
`;

exports[`testing ExpenseList component with values 1`] = `
<div>
  <div>
    <ExpenseListItem
      amount={50}
      createdAt={"1970-01-01T00:00:00.000Z"}
      description="laundry"
      id="1"
      key="list1"
      note=""
    />
  </div>
  <div>
    <ExpenseListItem
      amount={910}
      createdAt={86400000}
      description="rent"
      id="2"
      key="list2"
      note=""
    />
  </div>
  <div>
    <ExpenseListItem
      amount={800}
      createdAt={691200000}
      description="food"
      id="3"
      key="list3"
      note=""
    />
  </div>
  <div>
    <ExpenseListItem
      amount={110}
      createdAt={2419200000}
      description="mobile"
      id="4"
      key="list4"
      note=""
    />
  </div>
  <div>
    <ExpenseListItem
      amount={25}
      createdAt={-345600000}
      description="gym"
      id="5"
      key="list5"
      note=""
    />
  </div>
</div>
`;
